sw/host/ot_certs/src/asn1/mod.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{Context, Result};
+
+pub mod builder;
+pub mod der;
+
+/// An ASN1 tag.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Tag {
+    // Universal tags.
+    Boolean,
+    BitString,
+    GeneralizedTime,
+    Integer,
+    OctetString,
+    Oid,
+    PrintableString,
+    Sequence,
+    Set,
+    Utf8String,
+    /// Context tags.
+    Context {
+        constructed: bool,
+        value: usize,
+    },
+}
+
+/// An ASN1 object identifier.
+#[derive(Debug, Clone, PartialEq, Eq, strum::Display)]
+pub enum Oid {
+    // Custom oid.
+    Custom(String),
+}
+
+impl Oid {
+    /// Return the standard notation of the OID as string.
+    pub fn oid(&self) -> &str {
+        match self {
+            Oid::Custom(oid) => oid,
+        }
+    }
+
+    /// Return the DER encoding of the OID as a list of bytes.
+    pub fn to_der(&self) -> Result<Vec<u8>> {
+        let mut components = self
+            .oid()
+            .split('.')
+            .map(|s| {
+                s.parse::<u32>()
+                    .with_context(|| format!("invalid OID component '{s}'"))
+            })
+            .collect::<Result<Vec<_>>>()?;
+        // From X.690 spec, section 8.19:
+        // The number of subidentifiers (N) shall be one less than the number of object identifier components
+        // in the object identifier value being encoded. The numerical value of the first subidentifier is derived
+        // from the values of the first two object identifier components in the object identifier value being encoded,
+        // using the formula: (X*40) + Y where X is the value of the first object identifier component and Y is the value
+        // of the second object identifier component.
+        components.reverse();
+        let first = components
+            .pop()
+            .expect("cannot call push_oid with an empty OID");
+        let second = components.pop().unwrap_or(0);
```

Comment:
Should this be `unwrap_or` or `expect`?

Reply:
It's correct: if ever you had an OID with a single component, then you want 0. This is an extremely unlikely occurrence though.